spinn_utilities/ranged/ranged_list.py

```diff
@@ -0,0 +1,513 @@
+
+        # If non-range-based, just go through the values
+        else:
+            for value in self._ranges[slice_start: slice_stop]:
+                yield value
+
+    def iter_ranges(self):
+        """
+        Fast NOT update safe iterator of the ranges
+
+        :return: yields each range one by one
+        """
+        # If range based just yield the ranges
+        if self._ranged_based:
+            for r in self._ranges:
+                yield r
+
+        # If non-range based, build the ranges
+        else:
+            previous_value = self._ranges[0]
+            previous_start = 0
+            for start, value in enumerate(self._ranges):
+                if value != previous_value:
+                    yield (previous_start, start, previous_value)
+                    previous_start = start
+                    previous_value = value
+            yield (previous_start, start + 1, value)
+
+    def iter_ranges_by_slice(self, slice_start, slice_stop):
+        """
+         Fast NOT update safe iterator of the ranges covered by this slice
+
+         Note: The start and stop of the range will be reduced to just the
+         ids inside the slice
+
+         :return: yields each range one by one
+         """
+        slice_start, slice_stop = self._check_slice_in_range(
+            slice_start, slice_stop)
+
+        # If range-based, go through ranges that intersect the slice
+        if self._ranged_based:
+            for (start, stop, value) in self._ranges:
+                if slice_start < stop:
+
+                    # The range is updated so that the start and stop values
+                    # are within the slice requested
+                    yield (max(start, slice_start), min(stop, slice_stop),
+                           value)
+                    if slice_stop <= stop:
+                        break
+
+        # If non-range based, just go through the values
+        else:
+            for index, value in \
+                    enumerate(self._ranges[slice_start: slice_stop]):
+                yield (slice_start + index, slice_start + index + 1, value)
+
+    @staticmethod
+    def _is_list(value, size):
+        """ Determines if the value is a list of a given size.
+            An exception is raised if value *is* a list but is shorter\
+            than size
+        """
+
+        # If the item is a list, check the size is valid
+        if hasattr(value, '__iter__'):
+            if len(value) != size:
+                raise Exception(
+                    "The number of values does not equal the size")
+            else:
+                return True
+
+        # If the item has a "next" method with an argument "n" it is assumed
+        # that the value of size can be passed as the argument "n" to create
+        # a list of the correct size
+        if hasattr(value, 'next'):
+            arg_spec = inspect.getargspec(value.next)
+            if "n" in arg_spec.args:
+                return True
```

Comment:
I can see why this has been done (for support of RandomDistribution in PyNN), but this seems specific to this one example.  For now, it might make more sense to do this up front in sPyNNaker and then the value here *is* just a list.

A future interface would be to support values which are "Expandable" i.e. you set the values of the range to an object that extends ExpandableValue.  This is then a single range entry, but when you iterate over the range, if the value is an ExpandableValue, you iterate over its values in turn.  The RandomDistribution can then be wrapped in an ExpandableValue, which might include working with the seed of the distribution to avoid storing values.

Reply:
I chose not to do this as the RandomDistribution is passed in in various different ways and that would require checking for it everywhere.

Reply:
I understand this, but this makes the module dependent on PyNN, even if not explicitly so (and the implicit dependence is even more difficult to track, especially if PyNN changes something).  Additionally, this makes it hard to support future changes which may avoid expanding the RandomDistribution object, since at some point you will need to detect that you have one, even if it is going to be wrapped up in another object.

My suggestion would therefore be:
- Remove this check from here
- Add a create_list_function as an argument to the constructor of the RangeDictionary, with a default value of RangedList.  The documentation of this parameter should indicate that the call should expect the same arguments as the RangedList constructor expects, and should return an AbstractList. Replace direct calls to RangeList with calls to this function.
- In sPyNNaker, the RangedList can be subclassed and some methods can be overridden to detect and expand the RandomDistribution.  The calls that need to be overridden are:
    - set_value
    - set_value_by_slice
    - set_value_by_ids
- In sPyNNaker, the RangeDictionary can be subclassed with a constructor that simply passes in the RangedList subclass as the create_list_function method.

I am happy to make these changes myself if agreed.

Reply:
It can actually be done much easier that that.

Create a list_factory method in range_dictionary.py which by default calls the RangeList(....)

Then in spynakker add two classes SpynakkerRangedDict and SynakkerRandedList  

SpynakkerRangedDict just overwrites  list_factory method to call SynakkerRandedList()

 SynakkerRandedList can then for now over write _is_list and _as_list to check for and use pynn Random.  All set method use these two helpers.

Later if we want to do random on the fly  SynakkerRandedList can be changed accordingly.

That sound OK?

Reply:
Yes, this is just as good; I was avoiding overriding the "private" methods but they could be considered "protected" instead...

Reply:
The two methods are stateless support methods so can easily be made none "private."

Reply:
I don't mind either way - Python has no "protected" so there isn't a good way to do this!  One thing to be careful of is that I made them static - you might have to undo that to make it work correctly...

Reply:
That they are static is actually nicer as it allows a clean call to RangeList.asList(...)

I checked all the usages where still self.asList()

Done!

This is tested in https://github.com/SpiNNakerManchester/sPyNNaker8/pull/86